main.go

```diff
@@ -29,140 +31,153 @@ import (
+	cmd.Flags().AddGoFlagSet(flag.CommandLine) // make sure we get the klog flags
+	err := cmd.Flags().Parse(os.Args)
+	if err != nil {
+		logErrorAndExit(err, "Unable to parse flags")
+	}
+
+	adapterCfg, err := config.Parse()
+	if err != nil {
+		logErrorAndExit(err, "Unable to parse adapter configuration")
+	}
+
+	logger.Info("Starting monitoring server...")
+	monitoringServer := monitoring.NewServer(adapterCfg.MetricServers, cmd.MonitoringPort, adapterCfg.ReadinessProbe.FailureThreshold)
+	go monitoringServer.Start()
+
+	apmTracer, err := apm.NewTracer(serviceType, serviceVersion)
+	if err != nil {
+		logErrorAndExit(err, "Unable to create APM tracer")
+	}
+	apmTracer.SetLogger(&tracing.Logger{})
+
+	metricsClients, err := cmd.newMetricsClients(adapterCfg, apmTracer)
+	if err != nil {
+		logErrorAndExit(err, "Unable to create metrics provider")
+	}
+
+	scheduler := scheduler.NewScheduler(metricsClients...)
+	metricsRegistry := registry.NewRegistry()
+	scheduler.
+		WithMetricListeners(monitoringServer, metricsRegistry).
+		WithErrorListeners(monitoringServer).
+		Start().
+		WaitInitialSync()
+	aggProvider := provider.NewAggregationProvider(metricsRegistry, apmTracer)
+
+	cmd.WithCustomMetrics(aggProvider)
+	cmd.WithExternalMetrics(aggProvider)
+	if cmd.Insecure {
+		cmd.Authentication = nil
+		cmd.Authorization = nil
+	}
+
+	logger.Info("Starting elastic k8s metrics adapter...")
+	if err := cmd.Run(wait.NeverStop); err != nil {
+		logErrorAndExit(err, "Unable to run elastic k8s metrics adapter")
+	}
+}
+
 type ElasticsearchAdapter struct {
 	basecmd.AdapterBase
-	monitoringServer *monitoring.Server
 
 	Insecure                 bool
 	PrometheusMetricsEnabled bool
 	MonitoringPort           int
 }
 
-func (a *ElasticsearchAdapter) makeProviderOrDie(adapterCfg *config.Config) cm_provider.MetricsProvider {
+func (a *ElasticsearchAdapter) newMetricsClients(adapterCfg *config.Config, tracer *apm.Tracer) ([]client.Interface, error) {
 	dynamicClient, err := a.DynamicClient()
 	if err != nil {
-		klog.Fatalf("unable to construct dynamic dynamicClient: %v", err)
+		return nil, err
 	}
 
 	mapper, err := a.RESTMapper()
 	if err != nil {
-		klog.Fatalf("unable to construct dynamicClient REST mapper: %v", err)
+		return nil, err
 	}
 
-	tracer := createTracer()
-
 	var clients []client.Interface
 	for _, clientCfg := range adapterCfg.MetricServers {
 		switch clientCfg.ServerType {
-		case "elasticsearch":
+		case elastisearchMetricServerType:
 			esMetricClient, err := elasticsearch.NewElasticsearchClient(
 				clientCfg,
 				dynamicClient,
 				mapper,
 				tracer,
 			)
 			if err != nil {
-				klog.Fatalf("unable to construct Elasticsearch dynamicClient: %v", err)
+				return nil, err
 			}
 			clients = append(clients, esMetricClient)
-		case "custom":
+		case customMetricServerType:
 			kubeClientCfg, err := a.ClientConfig()
 			if err != nil {
-				klog.Fatalf("unable to construct Kubernetes dynamicClient config: %v", err)
```

Comment:
Should we keep the additional messages to give more context in case of error?

Reply:
Yes, I was a little hasty in simplifying, that might be useful.
Additional messages restored in 5e67bc0.